### ISSUE TITLE
Problem with StepSmoother

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1335]
**Modified by dietmarw on 25 Oct 2013 20:09 UTC**
In MSL 3.2.1 the Modelica.Fluid.Dissipation.Utilities.Functions.General.Stepsmoother function is written as follows:
```mo
        function Stepsmoother "Continuous interpolation for x "

          extends Modelica.Icons.Function;
          input Real func "input value for that result = 100%";
          input Real nofunc "input value for that result = 0%";
          input Real x "input variable for continuous interpolation";
          output Real result "output value";

        protected
          Real m=Modelica.Constants.pi/(func - nofunc);
          Real b=-Modelica.Constants.pi/2 - m*nofunc;
          Real r_1=tan(m*x + b);

        algorithm
          result := if x >= 0.999999*(func - nofunc) + nofunc and func > nofunc or x
             <= 0.999999*(func - nofunc) + nofunc and nofunc > func then 1 else if x
             <= 0.000001*(func - nofunc) + nofunc and func > nofunc or x >= 0.000001*(
            func - nofunc) + nofunc and nofunc > func then 0 else ((0.5*(exp(r_1) - exp(
            -r_1))/(0.5*(exp(r_1) + exp(-r_1))) + 1)/2);
          annotation (
            derivative=Stepsmoother_der,
            Documentation(info="<html> ... </html>"),  smoothOrder=5);
        end Stepsmoother;
```
The problem is that the value of r_1 is computed unnecessarily, and in cases where it might introduce a NaN. In addition, the final clause of the if can be written much more efficiently.

These two problems can be easily corrected by re-writing the function as follows:
```mo
        function Stepsmoother "Continuous interpolation for x "

          extends Modelica.Icons.Function;
          input Real func "input value for that result = 100%";
          input Real nofunc "input value for that result = 0%";
          input Real x "input variable for continuous interpolation";
          output Real result "output value";

        protected
          Real m=Modelica.Constants.pi/(func - nofunc);
          Real b=-Modelica.Constants.pi/2 - m*nofunc;

        algorithm
          result := if x >= 0.999999*(func - nofunc) + nofunc and func > nofunc or x
             <= 0.999999*(func - nofunc) + nofunc and nofunc > func then 1 else if x
             <= 0.000001*(func - nofunc) + nofunc and func > nofunc or x >= 0.000001*(
            func - nofunc) + nofunc and nofunc > func then 0 else
            (1-Modelica.Math.tanh(Modelica.Math.tan(m*x + b)))/2);
          annotation (
            derivative=Stepsmoother_der,
            Documentation(info="<html> ... </html>"),  smoothOrder=5);
        end Stepsmoother;
```
Note that this problem is not present in Stepsmoother_der, the derivative of this function. In that case the efficient form is already in use.

----

**Reported by awittkop on 25 Oct 2013 19:48 UTC**
In MSL 3.2.1 the Modelica.Fluid.Dissipation.Utilities.Functions.General.Stepsmoother function is written as follows:

        function Stepsmoother "Continuous interpolation for x "

          extends Modelica.Icons.Function;
          input Real func "input value for that result = 100%";
          input Real nofunc "input value for that result = 0%";
          input Real x "input variable for continuous interpolation";
          output Real result "output value";

        protected
          Real m=Modelica.Constants.pi/(func - nofunc);
          Real b=-Modelica.Constants.pi/2 - m\*nofunc;
          Real r_1=tan(m\*x + b);

        algorithm
          result := if x >= 0.999999\*(func - nofunc) + nofunc and func > nofunc or x
             <= 0.999999\*(func - nofunc) + nofunc and nofunc > func then 1 else if x
             <= 0.000001\*(func - nofunc) + nofunc and func > nofunc or x >= 0.000001\*(
            func - nofunc) + nofunc and nofunc > func then 0 else ((0.5\*(exp(r_1) - exp(
            -r_1))/(0.5\*(exp(r_1) + exp(-r_1))) + 1)/2);
          annotation (
            derivative=Stepsmoother_der,
            Documentation(info="<html> ... </html>"),  smoothOrder=5);
        end Stepsmoother;

The problem is that the value of r_1 is computed unnecessarily, and in cases where it might introduce a NaN. In addition, the final clause of the if can be written much more efficiently.

These two problems can be easily corrected by re-writing the function as follows:

        function Stepsmoother "Continuous interpolation for x "

          extends Modelica.Icons.Function;
          input Real func "input value for that result = 100%";
          input Real nofunc "input value for that result = 0%";
          input Real x "input variable for continuous interpolation";
          output Real result "output value";

        protected
          Real m=Modelica.Constants.pi/(func - nofunc);
          Real b=-Modelica.Constants.pi/2 - m\*nofunc;

        algorithm
          result := if x >= 0.999999\*(func - nofunc) + nofunc and func > nofunc or x
             <= 0.999999\*(func - nofunc) + nofunc and nofunc > func then 1 else if x
             <= 0.000001\*(func - nofunc) + nofunc and func > nofunc or x >= 0.000001\*(
            func - nofunc) + nofunc and nofunc > func then 0 else
            (1-Modelica.Math.tanh(Modelica.Math.tan(m\*x + b)))/2);
          annotation (
            derivative=Stepsmoother_der,
            Documentation(info="<html> ... </html>"),  smoothOrder=5);
        end Stepsmoother;

Note that this problem is not present in Stepsmoother_der, the derivative of this function. In that case the efficient form is already in use.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1335

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 25 Oct 2013 20:09 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by wischhusen on 28 Oct 2013 07:59 UTC**
Changed in b889aaff2094fece7aead53d5ca23bcdbe338f57.

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 30 Oct 2013 15:35 UTC**
My apologies, I made a sign error in the StepSmoother function rewrite.

The line:
```mo
(1-Modelica.Math.tanh(Modelica.Math.tan(m*x + b)))/2);
```
should read:
```mo
(1+Modelica.Math.tanh(Modelica.Math.tan(m*x + b)))/2);
```
Could this be fixed?

**Comment by modelica-trac-importer** [comment]
**Comment by wischhusen on 30 Oct 2013 16:03 UTC**
Done with 4dfa0041397571afec0dc8b836578298d7e8dbf4.

**Comment by modelica-trac-importer** [comment]
**Modified by wischhusen on 30 Apr 2014 10:12 UTC**

**Comment by modelica-trac-importer** [comment]
**Changelog modified by wischhusen on 30 Apr 2014 10:12 UTC**
Corrected Stepsmoother function to avoid numerical problems

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 30 Apr 2014 10:44 UTC**
Actually this was only corrected in trunk which is the next MSL 3.2.2. For 3.2.1+build.3 it also needs to be applied to the maintenance branch.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 30 Apr 2014 12:21 UTC**
Maintenance actually fixed in 84c6a6b2a36033aeac790aac0f67888efcdaf483, ca5f96f354a4f3bcd137b5a3314859dd7b473e5a but was forgotten to mention here when closing.
